### DISCUSSION TITLE
How to redirect all pages to /login if no user found in session? Looking for examples.

### DISCUSSION THREAD

**Discussion by Dannymx** [vercel/next.js#14945] [Help]
Is there a good example (maybe from Next.js examples) that shows how to redirect all pages to a `/login` page if no user found in the session?

I checked some examples but the `with-iron-session` redirects to `/login` from within the page, I'm looking for a more global way to redirect and maybe opt-out pages (ie. About us) that don't need authentication.

I also see the `with-firebase-authentication` is also using something similar at page level:
```
const Index = () => {
  const { user, logout } = useUser()
```

There the `userUser` hook will redirect in case not user is found. 

Would this possible to do, let's say with the `userHook` being called in `_app.js` and redirect all pages from there when no user is found? And possibly pass a param from an `About us` page so the auth check can be skipped?

I'm just looking for ideas 👍

**Comment by mattcarlotta** [comment]
Assuming you're not using a custom server and you're using context or redux and dealing with page-level authentication, then there really isn't a one-size-fits-all/great way to handle global auth checks if you want to keep site-wide ASO (automatic static optimization) and quick responses/interaction. 

For reference, [Web vitals](https://nextjs.org/docs/advanced-features/measuring-performance):
TTFB - Time To First Byte
FCP - First Contentful Paint
FID - First Input Delay

Current options are (the examples below include redux and cookie-session):

1.) Check in getInitialProps within a custom _app. 

This blocks the **entire** app from loading, opts out of ASO, and results in a slower TTFB/FCP. You'll also have to keep a blacklist of which routes require auth and which ones don't. This can become quite a nuisance when dealing with hundreds-thousands of dynamic routes.

**_app.js**
```js
import App from "next/app"
impoty get from "lodash.get";
import wrapper from "../store";

class MyApp extends App {
  static async getInitialProps({ Component, ctx }) {
    const { store: { getState }, req, res } = ctx;
     // get user role from redux state
    const role = getState(({ user }) => user.role); 
 
    if(!role || role === "guest") {
      try {
        // pull out cookie from "req.headers.cookie"
       const cookie = get(req, ["headers", "cookie"]); 
        // check if cookie exists 
       const config = cookie ? { headers: { cookie } } : undefined;
        // check if user is logged in
       const hasSession =  await axios.get("/user/signedin", config);
        // check if user has permissions to view page, otherwise throw Error
        ...authentication page logic...
      } catch (e) {
       if(res) {
           // if this is an initial page load, redirect on server-side
          res.setHeader(307, "/signin")
          res.end();
       } else {
          // else redirect client-side
         Router.push("/signin");       
       }
       return { };
    }
  }

  return {
    pageProps: {
      ...(Component.getInitialProps ?
        await Component.getInitialProps(ctx) :
        {}),
    },
  };
   
  render() {
     const { Component, pageProps } = this.props;
     return <Component { ...pageProps } />;
  }
}

export default wrapper.withRedux(MyApp);
```

2.) Check in page-level getServerSideProps/getStaticProps. 

Similar to the above, however, this will keep ASO and doesn't render block the entire app, but it'll still render block the **page** from loading (and result in a slower TTFB/FCP if using getServerSideProps). In addition, it'll need to be replicated across any authenticated pages, and as of writing is currently a [RFC](https://github.com/vercel/next.js/discussions/14890).

**Settings.js** (an authenticated page)
```js
import get from "lodash.get";
import wrapper from "../store";

const Settings = () => <h1>User Settings Page</h1>;

export const getServerSideProps = wrapper.getServerSideProps(async ctx => {
 const { store: { getState }, req } = ctx;
  // get user role from redux state
 const role = getState(({ user }) => user.role);
 
 if(!role || role === guest) {
   try {
     // pull out cookie from "req.headers.cookie"
    const cookie = get(req, ["headers", "cookie"]); 
     // check if cookie exists 
    const config = cookie ? { headers: { cookie } } : undefined;
     // check if user is logged in
    const hasSession =  await axios.get("/user/signedin", config);
     // check if user has permissions to view page, otherwise throw Error
    ...authentication page logic...
   } catch (e) {
     return {
      redirect: {
        permanent: false,
        destination: "/signin"
      }
    }
   }
 }

  return { props: {} };
});

export default Settings;
```

3.) Check in a reusable page-level HOC (higher-order component). 

This keeps ASO and results in a quicker TTFB/FCP because it doesn't render block any pages from loading, so you'll have to utilize some placeholders (like [skeletons](https://material-ui.com/components/skeleton/#wave-example) or [spinners](https://material-ui.com/components/progress/#circular-indeterminate)). It also checks for a logged-in session once and only once without any additional logic. The disadvantage of this is that your auth pages won't be SSR'd, nor can they [previewable/pre-rendered](https://nextjs.org/docs/advanced-features/preview-mode). It also requires an additional request (additional request other than the requested page) in a custom _app's useEffect lifecycle, and all redirects will happen on the **client-side only.** While this results in a quicker TTFB/FCP and is reusable for any other auth pages, it, unfortunately, results in a slower FID.

**_app.js**
```js
import { useDispatch, useSelector } from "react-redux";
import wrapper from "../store";

const App = ({ Component, pageProps }) => {
   // get user role from redux state
  const role = useSelector(({ auth }) => auth.role);
  const dispatch = useDispatch();
  
  useEffect(() => {
     // a redux action that sends a one-time request to "/user/signed" on initial page load
     // which sets a role as "guest" or "member" when resolved
     // since this is a client-side request, the cookie logic isn't needed
    if (!role) dispatch(signinOnLoad());
   }, []);

  return <Component { ...pageProps } />
};

```
Reusable **withAuthentication** HOC (like a middleware component):
```js
import { useSelector } from "react-redux";

const withAuthentication = WrappedComponent => {
  const RequiresAuthentication = props => {
     // get user role from redux state
   const role = useSelector(({ auth }) => auth.role);

   useEffect(() => {
     // if a there isn't a logged in user and their role has been set to "guest"
     // then redirect them to "/signin"
    if(role === "guest") Router.push("/signin");
   }, [role]);

     // if there's a loggedInUser, show the wrapped page, otherwise show a loading indicator
    return role && role !== "guest" ? <WrappedComponent {...props} /> : <div>Loading...</div>;
  };

  return RequiresAuthentication;
};

withAuthentication.propTypes = {
  WrappedComponent: PropTypes.node.isRequired
};

export default withAuthentication;
```

**Settings.js** (authenticated page)
```js
import withAuthentication from "../containers/withAuthentication";

const Settings = () => <h1>User Settings Page</h1>;

export default withAuthentication(Settings);
```

---

Until option 2 becomes available, I'd recommend option 3 (this is what I'm currently using). For simple authentication logic, option 1 is like taking a bulldozer to the entire app,  which really isn't necessary nor ideal. Option 1 also doesn't play well with redux since the state is initially empty during a page load. So if you set some state to redux on the server, then redirect on the server, then unfortunately the state will be reset again.

**Reply by jacksonblankenship** [reply]
Thanks for this. It's exactly what I've been looking for. 👍

**Reply by yelnyafacee** [reply]
hi, how do I make method 3 work with `next-auth`?

`next-auth` got its own `{ Provider } from 'next-auth/client'` will `redux` conflict with it?
